src/shared/sleep-config.c

```diff
@@ -207,67 +292,203 @@ static bool battery_discharge_rate_is_valid(int battery_discharge_rate) {
                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery discharge percentage rate per hour: %m");
 
-        r = get_system_battery_identifier_hash(&system_hash_id);
-        if (r < 0)
-               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
-
-        r = write_string_filef(
+        if (truncate)
+                r = write_string_filef(
                         DISCHARGE_RATE_FILEPATH,
-                        WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_MKDIR_0755,
+                        WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_MKDIR_0755 | WRITE_STRING_FILE_TRUNCATE,
                         "%"PRIu64" %d",
                         system_hash_id,
                         estimated_battery_discharge_rate);
+        else
+                r = write_string_filef(
+                        DISCHARGE_RATE_FILEPATH,
+                        WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_MKDIR_0755,
+                        "%"PRIu64" %d",
+                        system_hash_id,
+                        estimated_battery_discharge_rate);
+
         if (r < 0)
                 return log_debug_errno(r, "Failed to create %s: %m", DISCHARGE_RATE_FILEPATH);
 
         log_debug("Estimated discharge rate %d successfully updated to %s", estimated_battery_discharge_rate, DISCHARGE_RATE_FILEPATH);
+
+        return 0;
+}
+
+/* Estimate battery discharge rate using stored previous and current capacity over timestamp difference */
+int estimate_battery_discharge_rate_per_hour(
+                                        Hashmap *last_capacity,
+                                        Hashmap *current_capacity,
+                                        usec_t before_timestamp,
+                                        usec_t after_timestamp) {
+
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        bool truncate = true;
+        int battery_last_capacity, battery_current_capacity, battery_discharge_rate, r;
+
+        assert(last_capacity);
+        assert(current_capacity);
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name;
+                uint64_t system_hash_id;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name: %m");
+                        continue;
+                }
+
+                battery_last_capacity = PTR_TO_INT(hashmap_get(last_capacity, battery_name));
+                if (battery_last_capacity <= 0)
+                        continue;
+
+                battery_current_capacity = PTR_TO_INT(hashmap_get(current_capacity, battery_name));
+                if (battery_last_capacity <= 0)
```

Comment:
this should check battery_current_capacity

Reply:
sorry... missed in copy paste.fixed